### ISSUE TITLE
Optimization for string interpolation - simple scenarios: unfolding constants and lowering to concatenation

### ISSUE THREAD

**Issue by abonie** [dotnet/fsharp#16247]
Seems like there is a lot we could do to improve perf of string interpolation. One good example is https://github.com/fsharp/fslang-suggestions/issues/1108, but also some low hanging fruits to start with could be unfolding compile time constants, e.g.:
```fsharp
[<Literal>]
let x = "hello"

[<Literal>]
let y = "world"

let z = $"{x} {y}!"
```
in the above, the last line could be lowered to `let z = "hello world!"`

Another one could be lowering to `System.String.Concat` when it applies, e.g. `let f (x: string) (y: string) = $"{x},{y}."` could be rewritten as `let f (x: string) (y: string) = System.String.Concat(x, ",", y, ".")`, because we know that all expression holes are filled with objects of type string and there are at most 4 pieces to concatenate.

**Comment by abonie** [comment]
For the second one, here is difference in allocations from calling the first version of the function vs. the second one 1k times in a loop. Using concat:
![image](https://github.com/dotnet/fsharp/assets/20281641/9923f894-7993-4206-b208-c7ae26b06f1e)
vs. string interpolation:
![image](https://github.com/dotnet/fsharp/assets/20281641/91424b1b-87cd-4139-80d5-b43b4d86f329)

**Comment by charlesroddie** [comment]
@abonie I see that the PR only works when the inner expressions are strings only. That is not the majority case. Will the PR make it easier to make all expressions concats (i.e. is it a step towards using a collector for all interpolation https://github.com/fsharp/fslang-suggestions/issues/1108) or will it make this harder by adding an exception that needs to be maintained in the future?

**Comment by abonie** [comment]
> the PR only works when the inner expressions are strings only. That is not the majority case.

That is true. This issue is specifically about lowering to `System.String.Concat` and it doesn't really make sense for other cases (when inner exprs are not strings).

> Will the PR make it easier to make all expressions concats (i.e. is it a step towards using a collector for all interpolation https://github.com/fsharp/fslang-suggestions/issues/1108) or will it make this harder by adding an exception that needs to be maintained in the future?

I think neither - as far as I can tell, this is largely orthogonal to interpolated string handlers, doesn't make it particularly easier or more difficult to implement them and also won't be superseded by them, since for cases where this optimization applies, it should still be faster than using a collector.

Reopened the issue, because it also mentions unfolding constants, which the PR does not cover. However, will have to estimate how much of a difference that would make vs work needed. Would be nice to tackle interpolated string handlers soon.

**Comment by kerams** [comment]
So the only remaining actionable item is constant folding? Would love to be able to do `let [<Literal>] z = $"{x} {y}!"` since pure string interpolation is already lowered to concat today, and `x + " " + y + "!"` can be assigned to a literal.

Although funnily enough, `let z = x + " " + y + "!"` emits concat for some reason unless `[<Literal>]` is used.

**Comment by T-Gro** [comment]
Constant folding are the last "low hanging" element I believe, the next step after that would be interpolated string handlers.

Automatic derivation of constants (from `+` or from `$"{first} {second}"` ) that are not explicitely annotated will have some program size impact - even though noticeable only in degenerate situations like huge amounts of interpolated strings from a smaller set of atomic constants.


Should go via a language suggestion first if we would want to do that.
